src/ReinforcementLearningCore/src/policies/agent.jl

```diff
@@ -13,21 +54,27 @@ is a Callable and its call method accepts varargs and keyword arguments to be
 passed to the policy. 
 
 """
-mutable struct Agent{P,T} <: AbstractPolicy
+mutable struct Agent{P,T,C} <: AbstractPolicy
     policy::P
     trajectory::T
-    cache::NamedTuple # trajectory do not support partial inserting
+    cache::C # trajectory does not support partial inserting
+
+    function Agent(policy::P, trajectory::T; env=missing) where {P,T}
+        if !ismissing(env)
+            cache = AgentCache(policy, env)
+            agent = new{P,T, typeof(cache)}(policy, trajectory, cache)
+        else
+            agent = new{P,T, typeof(AgentCache())}(policy, trajectory, AgentCache())
```

Comment:
Another possibility is to use the trajectory to infer the type of the cache's state and action types. By default a trajectory will assume they are Integers. But if you have Vector states you can specify it when constructing the trajectory.